ouroboros-network/src/Ouroboros/Network/RecentTxIds.hs

```diff
@@ -0,0 +1,152 @@
+
+{-----------------------------------------------------------------------------
+  Construction
+-----------------------------------------------------------------------------}
+
+-- | \( O(1) \). An empty 'RecentTxIds'.
+empty :: RecentTxIds txid
+empty = RecentTxIds OrdPSQ.empty
+
+{-----------------------------------------------------------------------------
+  Insertion
+-----------------------------------------------------------------------------}
+
+-- | \( O(\log(n)) \). Insert a new transaction ID and 'Time' (expected to
+-- represent the current time) into the 'RecentTxIds'.
+insertTxId :: Ord txid => txid -> Time -> RecentTxIds txid -> RecentTxIds txid
+insertTxId txId time (RecentTxIds psq) = RecentTxIds $
+  OrdPSQ.insert txId time () psq
+
+-- | \( O(m*\log(n)) \). Insert new transaction IDs and 'Time' (expected to
+-- represent the current time) into the 'RecentTxIds'.
+insertTxIds :: (Foldable t, Ord txid)
+            => t txid
+            -> Time
+            -> RecentTxIds txid
+            -> RecentTxIds txid
+insertTxIds txIds time txPsq = foldl'
+  (\acc txId -> insertTxId txId time acc)
+  txPsq
+  txIds
+
+{-----------------------------------------------------------------------------
+  Expiration
+-----------------------------------------------------------------------------}
+
+-- | The time threshold after which transaction IDs should expire.
+newtype ExpiryThreshold = ExpiryThreshold DiffTime
+
+-- | Expire/remove transaction IDs that are associated with expiry 'Time's
+-- less than or equal to the provided 'Time'.
+--
+-- The 'fst' of the result contains those transaction IDs that are expired
+-- while the 'snd' of the result contains the rest of the 'RecentTxIds' stripped
+-- of those transaction IDs.
+expireTxIds
+  :: Ord txid
+  => ExpiryThreshold
+  -> Time  -- ^ The current time
+  -> RecentTxIds txid
+  -> ([(txid, Time)], RecentTxIds txid)
+expireTxIds (ExpiryThreshold expiryThreshold) currentTime (RecentTxIds psq) =
+    (expired', RecentTxIds psq')
+  where
+    expireOlderThan :: Time
+    expireOlderThan = Time (currentTime `diffTime` Time expiryThreshold)
+
+    (expired, psq') = OrdPSQ.atMostView expireOlderThan psq
+    expired' = [(txId, time') | (txId, time', _) <- expired]
+
+{-----------------------------------------------------------------------------
+  Conversion
+-----------------------------------------------------------------------------}
+
+-- | \( O(n) \). Convert to a list of pairs of transaction IDs and their
+-- expiry times. The order of the list is not specified.
+toList :: RecentTxIds txid -> [(txid, Time)]
+toList (RecentTxIds psq) = OrdPSQ.fold'
+  (\k p _ acc -> (k, p):acc)
+  []
+  psq
+
+{-----------------------------------------------------------------------------
+  Orphan instances
+-----------------------------------------------------------------------------}
+
+deriving via OnlyCheckIsWHNF "OrdPSQ" (OrdPSQ k p v)
+  instance NoUnexpectedThunks (OrdPSQ k p v)
```

Comment:
Do we need this orphaned instance? Why not define`OnlyCheckIsWHNF` instance directly for `RecentTxIds`?

Reply:
Actually, let's put a proper instance next to the one for `IntPSQ` here: https://github.com/input-output-hk/ouroboros-network/blob/64006b7331b3f70580ee70fad7808e7a0fe242d2/ouroboros-consensus/src/Ouroboros/Consensus/Util/Orphans.hs#L101

Reply:
@mrBliss: That instance for `IntPSQ` is in `ouroboros-consensus`. I could add a `Ouroboros.Network.Util.Orphans` module or, as Marcin suggested, define the instance for `RecentTxIds`.